packages/lit-dev-server/src/middleware/content-security-policy-middleware.ts

```diff
@@ -0,0 +1,120 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import type Koa from 'koa';
+
+/**
+ * Options for contentSecurityPolicyMiddleware.
+ */
+export interface ContentSecurityPolicyMiddlewareOptions {
+  /**
+   * Enables some additional directives that are only required when running in
+   * local dev mode.
+   */
+  devMode?: boolean;
+
+  /**
+   * If true, CSP violations will be reported to the Google CSP Collector.
+   */
+  reportViolations?: boolean;
+}
+
+/**
+ * Creates a Koa middleware that sets the lit.dev Content Security Policy (CSP)
+ * headers.
+ *
+ * Some good resources about CSP:
+ * https://web.dev/strict-csp/
+ * https://www.w3.org/TR/CSP3/
+ * https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP
+ * https://speakerdeck.com/lweichselbaum/csp-a-successful-mess-between-hardening-and-mitigation
+ */
+export const contentSecurityPolicyMiddleware = (
+  opts: ContentSecurityPolicyMiddlewareOptions = {}
+): Koa.Middleware => {
+  const cspHeaderValue = [
+    // TODO(aomarks) We should also enable trusted types, but that will require
+    // a policy in playground-elements for creating the worker, and a policy
+    // es-module-lexer for doing an eval (see next comment for more on that).
+
+    // TODO(aomarks) unsafe-inline is needed because we use some inline scripts
+    // on some pages. We should generate hashes for them during Eleventy build
+    // and allowlist them here using `<hash-algorithm>-<base64-value>`
+    // directives.
+    //
+    // TODO(aomarks) unsafe-eval is needed for an eval that is made by
+    // es-module-lexer to perform JavaScript string unescaping
+    // (https://github.com/guybedford/es-module-lexer/blob/91964da6b086dc5029091eeef481180a814ce24a/src/lexer.js#L32).
+    // There are a number of ways we could make this stricter: [1] use the
+    // asm.js build which doesn't use eval (but it's significantly slower), [2]
+    // use a separate CSP for the worker (except Chrome doesn't support that
+    // yet, though we could simulate it with an iframe), [3] get trusted types
+    // into the WASM build, [4] modify or fork the WASM build to implement
+    // string unescaping without eval (needs benchmarking).
+    //
+    // In dev mode, data: scripts are required because @web/dev-server uses them
+    // for automatic reloads.
+    `script-src 'self' 'unsafe-inline' 'unsafe-eval' https://www.googletagmanager.com/gtag/js ${
+      opts.devMode ? ` data:` : ''
+    }`,
+
+    // unpkg.com is needed to allow the Playground worker to fetch dependencies.
```

Comment:
Does this mean that we don't support any other cdns (skypack, jsdelivr, etc.)?

Reply:
We always use unpkg in the Playground web worker, that's hard-coded. If you write a bare module in a Playground, it will be retrieved behind-the-scenes from unpkg.

However you can still directly import any arbitrary URL from a Playground editor -- because anything running in a Playground preview is unaffected by this CSP, because it runs on the totally separate playground.lit.dev origin.

Reply:
Is there anything that we (jsDelivr) can do to get added as well? We have support for ESM and a detailed API that even provides correct endpoints for each project https://github.com/jsdelivr/data.jsdelivr.com

Reply:
Hi @jimaek!

This CSP directive won't stop somebody using the lit.dev playground to import from jsDelivr if they chose to.  In general there is nothing you can do on the playground today that you won't be able to do after this policy is enforced. Here's an example that works [today](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy5qcyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnaHR0cHM6Ly9lc20ucnVuL2xpdCc7XG5cbmV4cG9ydCBjbGFzcyBTaW1wbGVHcmVldGluZyBleHRlbmRzIExpdEVsZW1lbnQge1xuICBzdGF0aWMgc3R5bGVzID0gY3NzYHAgeyBjb2xvcjogYmx1ZSB9YDtcblxuICBzdGF0aWMgcHJvcGVydGllcyA9IHtcbiAgICBuYW1lOiB7dHlwZTogU3RyaW5nfSxcbiAgfTtcblxuICBjb25zdHJ1Y3RvcigpIHtcbiAgICBzdXBlcigpO1xuICAgIHRoaXMubmFtZSA9ICdTb21lYm9keSc7XG4gIH1cblxuICByZW5kZXIoKSB7XG4gICAgcmV0dXJuIGh0bWxgPHA-SGVsbG8sICR7dGhpcy5uYW1lfSE8L3A-YDtcbiAgfVxufVxuY3VzdG9tRWxlbWVudHMuZGVmaW5lKCdzaW1wbGUtZ3JlZXRpbmcnLCBTaW1wbGVHcmVldGluZyk7XG4ifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPGhlYWQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vc2ltcGxlLWdyZWV0aW5nLmpzXCI-PC9zY3JpcHQ-XG48L2hlYWQ-XG48Ym9keT5cbiAgPHNpbXBsZS1ncmVldGluZyBuYW1lPVwiV29ybGRcIj48L3NpbXBsZS1ncmVldGluZz5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4yLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjEuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl4zLjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMi4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfV0), and it will work after this CSP policy is enforced too.

In [playground-elements](https://github.com/google/playground-elements), when you write a bare module specifier, we do our own bare module resolution, and fetch from unpkg behind the scenes. So this CSP directive is just to allow playground-elements to do that.

I do think it would be great to allow configuring which CDN is used, though. I previously filed https://github.com/google/playground-elements/issues/187 to track that. Please feel free to comment there!